### PR TITLE
Internal usage flags and rules

### LINKED ISSUE #296: Remove variations of "read-only" usages?
Given that buffer/texture usage belongs to an internal state of a resource (users don't set it explicitly), and we allow multiple read-only usages to co-exist, what is the point to distinguish between different read-only usages (from the specification point of view) at all?

Perhaps, we should write the specification in such a way that there is a single "Read" usage, and a resource can only be in one of the allowed usages at a time. I.e. the rules become simpler, and usage becomes purely exclusive.

For example, buffer usage rules would be rewritten as:
---
Buffers have the following usages (and commands inducing that usage):
  - `READ` for buffers that are read only by the following commands:
    - `WebGPUCommandEncoder.setVertexBuffers` parameter `buffers
    - `WebGPUCommandEncoder.setIndexBuffer` parameter `buffer`
    - `WebGPUCommandEncoder.{draw|drawIndexed|dispatch}Indirect` parameter `indirectBuffer`
    - `setBindGroup` parameter `bindGroup` created with one of the following types:
      - "uniform-buffer"
      - "dynamic-uniform-buffer"
    - `copyBufferToBuffer` parameter `src`
    - `copyBufferToTexture` parameter `source.buffer`
  - `STORAGE` for `setBindGroup` parameter `bindGroup` created with one the following types:
    - "storage-buffer"
    - "dynamic-storage-buffer"
  - `TRANSFER` for buffers used as the copy destination of `copy{Buffer,Texture}ToBuffer`

### LINKED ISSUE #745: Describe "shadow" usage bits
The usage scope validation considers usage bits at higher granularity than what the user specifies, as described in https://github.com/gpuweb/gpuweb/issues/514#issuecomment-623047982
We need to specify carefully what the "shadow" usage bits are (i.e. `STORAGE_READ`), and how they relate to the user-visible bits.

### PR DESCRIPTION
**Pull request by kvark**
Closes #745 
Closes #296
Likely conflicts with #972 (sorry!)

We describe the internal set of flags an implementation considers to track the resource usage, how they map to the descriptor flags provided by the user, and how they are validated. The "mutable"/"immutable" usage distinction is gone, and instead we have a list possible sets of compatible usages, as mentioned in https://github.com/gpuweb/gpuweb/pull/972#discussion_r468203784. It also includes clarifying the notion of a "subresource" as outlined in https://github.com/gpuweb/gpuweb/pull/972#discussion_r465373446

This PR needs more work on naming and editing, so that it looks better. Input is invited!

***
<a href="https://pr-preview.s3.amazonaws.com/kvark/gpuweb/pull/994.html" title="Last updated on Aug 19, 2020, 5:44 PM UTC (3be85e2)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/994/82d202e...kvark:3be85e2.html" title="Last updated on Aug 19, 2020, 5:44 PM UTC (3be85e2)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -587,107 +587,126 @@ the user behind an `ArrayBuffer`, thus avoiding any data copies.
...
+    : <dfn>copy-source</dfn>
+    ::  Source of a copy operation. Preserves the contents.
+        Allowed by {{GPUBufferUsage/COPY_SRC}} or {{GPUTextureUsage/COPY_SRC}}.
+    : <dfn>copy-destination</dfn>
+    ::  Destination of a copy operation.
+        Allowed by {{GPUBufferUsage/COPY_DST}} or {{GPUTextureUsage/COPY_DST}}.
+    : <dfn>input</dfn>
+    :: Buffer with input data for draw or dispatch calls. Preserves the contents.
+        Allowed by {{GPUBufferUsage/INDEX}}, {{GPUBufferUsage/VERTEX}}, or {{GPUBufferUsage/INDIRECT}}.
+    : <dfn>uniform</dfn>
```
Comment (Kangz) [inline spec/index.bs:601]: It's a bit weird that sampled is called uniform. Maybe we could bundle a bunch of usages as "shader read"? The problem is that we need storage to be separate :/

We might only need the following classes:

  - non-storage-readonly (horrible name though)
  - storage-readonly
  - storage

File: spec/index.bs [hunk 2]
```diff
@@ -587,107 +587,126 @@ the user behind an `ArrayBuffer`, thus avoiding any data copies.
...
-Enforcing this rule allows the API to limit when data races can occur
-when working with memory. That property makes applications written against
+Some [=internal usage=]s are compatible with others.
+That means an implementation can consider multiple usages active at the same time.
+There is a limited set of <dfn dfn>allowed usage set</dfn>, where each usage is compatible with the other:
+  - {[=copy-source=] | [=input=] | [=uniform=] | [=storage-read=]} - in any combination
+  - [=copy-destination=] - singular usage
+  - {[=storage=] | [=storage-read=]} - in any combination
+  - [=attachment=] - singular usage
+  - {[=attachment-read=] | [=uniform=]} - in any combination
```
Comment (Kangz) [inline spec/index.bs:638]: What about storage read?

File: spec/index.bs [hunk 3]
```diff
@@ -1446,10 +1465,10 @@ interface GPUBufferUsage {
                 - |this| is a [=valid=] {{GPUDevice}}.
                 - |descriptor|.{{GPUBufferDescriptor/usage}} is a subset of |this|.[[allowed buffer usages]].
                 - If |descriptor|.{{GPUBufferDescriptor/usage}} contains {{GPUBufferUsage/MAP_READ}}:
-                    - |descriptor|.{{GPUBufferDescriptor/usage}} contains no other [=usage flags=]
+                    - |descriptor|.{{GPUBufferDescriptor/usage}} is contains no other flags
```
Comment (Kangz) [inline spec/index.bs:1468]: nit: stray "is"?

File: spec/index.bs [hunk 4]
```diff
@@ -2720,41 +2735,36 @@ A {{GPUBindGroup}} object has the following internal slots:
...
+
+        <div algorithm>
+            <dfn abstract-op>derive binding usage</dfn>(type)
+
+            **Arguments:**
+                - {{GPUBindingType}} |type|
+
+            **Returns:** [=internal usage=]
+
+            1. If |type| is
```
Comment (kainino0x) [inline spec/index.bs:2758]: nit: Could optionally use `<dl class=switch>`

File: spec/index.bs [hunk 5]
```diff
@@ -3957,14 +3967,17 @@ dictionary GPUCommandEncoderDescriptor : GPUObjectDescriptorBase {
...
+                        is considered to be used as [=attachment=] for the
                         duration of the render pass.
                 1. Let |depthStencilAttachment| be |descriptor|.{{GPURenderPassDescriptor/depthStencilAttachment}}.
                 1. If |depthStencilAttachment| is not `null`:
-                    1. The texture [=subresource=] seen by |depthStencilAttachment|.{{GPURenderPassDepthStencilAttachmentDescriptor/attachment}}
-                        is considered to be used as {{GPUTextureUsage/OUTPUT_ATTACHMENT}} for the
-                        duration of the render pass.
+                    1. if |depthStencilAttachment|.{{GPURenderPassDepthStencilAttachmentDescriptor/depthReadOnly}} and
+                        {{GPURenderPassDepthStencilAttachmentDescriptor/stencilReadOnly}} are set
+                        1. The [=texture subresource=] seen by |depthStencilAttachment|.{{GPURenderPassDepthStencilAttachmentDescriptor/attachment}}
```
Comment (kainino0x) [inline spec/index.bs:3977]: nit: Texture subresources, plural, if we're tracking depth and stencil as separate subresources

File: spec/index.bs [hunk 6]
```diff
@@ -2595,14 +2614,10 @@ A {{GPUBindGroup}} object has the following internal slots:
     ::
         The set of {{GPUBindGroupEntry}}s this {{GPUBindGroup}} describes.
 
-    : <dfn>\[[usedBuffers]]</dfn> of type maplike<{{GPUBuffer}}, {{GPUBufferUsageFlags}}>.
+    : <dfn>\[[usedResources]]</dfn> of type maplike<[=subresource=], setLike<[=internal usage=]>>.
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply spec/index.bs:2617]: I guess maplike comes from WebIDL, but I don't think this is really an appropriate place to use it. I'd be fine with "map" and "set" I guess.

File: spec/index.bs (hunk 2 above)
Comment (kvark) [reply spec/index.bs]: I'm not sure if this is actually allowed :/  I guess, depending on https://github.com/KhronosGroup/Vulkan-Docs/issues/1345 we'll either allow storage-read here, or scrap the whole RODS out of WebGPU (or make it an extension?). So there is no harm in specifying here that storage-read is allowed, until we know more.

File: spec/index.bs [hunk 7]
```diff
@@ -587,107 +587,114 @@ the user behind an `ArrayBuffer`, thus avoiding any data copies.
...
-  - a single [=mutating usage=]
+We define <dfn dfn>subresource</dfn> to be either a whole buffer, or a [=texture subresource=].
+
+Some [=internal usage=]s are compatible with others.
+That means an implementation can consider multiple usages active at the same time.
+There is a limited set of <dfn dfn>allowed usage set</dfn>s, where each usage is compatible with the other:
+  - {[=internal usage/input=] | [=internal usage/constant=] | [=internal usage/storage-read=]} - in any combination
+  - {[=internal usage/storage=] | [=internal usage/storage-read=]} - in any combination
+  - [=internal usage/attachment=] - singular usage
+  - {[=internal usage/attachment-read=] | [=internal usage/storage-read=] | [=internal usage/constant=]} - in any combination
```
Comment (Kangz) [inline spec/index.bs:626]: Why not bundle this line with the first one? They seem to be split just because one talks about buffers and on about textures?

File: spec/index.bs [hunk 8]
```diff
@@ -587,107 +587,114 @@ the user behind an `ArrayBuffer`, thus avoiding any data copies.
...
-at any given time can only be in either:
-  - a combination of [=read-only usage=]s
-  - a single [=mutating usage=]
+We define <dfn dfn>subresource</dfn> to be either a whole buffer, or a [=texture subresource=].
+
+Some [=internal usage=]s are compatible with others.
+That means an implementation can consider multiple usages active at the same time.
+There is a limited set of <dfn dfn>allowed usage set</dfn>s, where each usage is compatible with the other:
+  - {[=internal usage/input=] | [=internal usage/constant=] | [=internal usage/storage-read=]} - in any combination
+  - {[=internal usage/storage=] | [=internal usage/storage-read=]} - in any combination
```
Comment (Kangz) [inline spec/index.bs:624]: For buffers it makes sense, but do we really want to have `writeonly-storage` and `readonly-storage` usable in the same scope for textures?

Comment (kvark) [reply spec/index.bs:624]: We talked about this a bit recently with Kai. It would be weird to forbid this, because the restriction is very arbitrary.
Somebody may hit this and wonder why we put the restriction in.

File: spec/index.bs (hunk 7 above)
Comment (kvark) [reply spec/index.bs:626]: Yeah, we can merge them. But it kinda nice to have separate cases for buffers and textures...

Comment (kainino0x) [reply spec/index.bs:626]: I think it's good to have them separate for buffers and textures. Maybe the two lists of sets should actually be separate: `a buffer binding must be in one of these sets: (...), a texture binding must be in one of these sets: (...)`

File: spec/index.bs (hunk 8 above)
Comment (kainino0x) [reply spec/index.bs:624]: My reasoning is that `readonly-storage-buffer` should always be strictly "better" (less restrictive) than `storage-buffer`. That is, if a storage buffer is only used as readonly, then changing the binding type to be readonly shouldn't break anything. But if it's not compatible with `storage-buffer` then it could break usage compatibility with some other `storage-buffer` usage in the scope.

File: spec/index.bs [hunk 9]
```diff
@@ -3156,23 +3154,29 @@ A {{GPUProgrammableStageDescriptor}} describes the entry point in the user-provi
...
+              :: the |binding| has to be a read-only storage texture with format of |entry|.{{GPUBindGroupLayoutEntry/storageTextureFormat}}
+              : {{GPUBindingType/"writeonly-storage-texture"}}
+              :: the |binding| has to be a writable storage texture with format of |entry|.{{GPUBindGroupLayoutEntry/storageTextureFormat}}
+              : {{GPUBindingType/"uniform-buffer"}}
+              :: the |binding| has to be a uniform buffer
+              : {{GPUBindingType/"storage-buffer"}}
+              :: the |binding| has to be a storage buffer
+              : {{GPUBindingType/"readonly-storage-buffer"}}
+              :: the |binding| has to be a read-only storage buffer
+
```
Comment (kainino0x) [inline spec/index.bs:3178]: nit: 4-space indent should avoid the need for this and other blank lines in this block

File: spec/index.bs [hunk 10]
```diff
@@ -3156,23 +3154,29 @@ A {{GPUProgrammableStageDescriptor}} describes the entry point in the user-provi
...
-            the |binding| has to be a uniform buffer.
-        1. If |entry|.{{GPUBindGroupLayoutEntry/type}} is {{GPUBindingType/"storage-buffer"}},
-            the |binding| has to be a storage buffer.
-        1. If |entry|.{{GPUBindGroupLayoutEntry/type}} is {{GPUBindingType/"readonly-storage-buffer"}},
-            the |binding| has to be a read-only storage buffer.
+        1. If |entry|.{{GPUBindGroupLayoutEntry/type}} is:
+            <dl class=switch>
+
+              : {{GPUBindingType/"sampler"}}
+              :: the |binding| has to be a non-comparison sampler
```
Comment (kainino0x) [inline spec/index.bs:3161]: "has to be" should be "is" to match the other style. (or should we change style and make it "must be"?)

File: spec/index.bs [hunk 11]
```diff
@@ -2720,41 +2723,36 @@ A {{GPUBindGroup}} object has the following internal slots:
...
+                        1. Let |internalUsage| be the result of [$derive binding usage$](|layoutBinding|.{{GPUBindGroupLayoutEntry/type}})
+                        1. Each [=subresource=] seen by |resource| is added to {{GPUBindGroup/[[usedResources]]}} as |internalUsage|.
                 </div>
             1. Return |bindGroup|.
 
             Issue: define the "effective buffer binding size" separately.
         </div>
+
+        <div algorithm>
+            <dfn abstract-op>derive binding usage</dfn>(type)
```
Comment (kainino0x) [inline spec/index.bs:2739]: nit:
```suggestion
            <dfn abstract-op>derive binding usage</dfn>(|type|)
```

**Comment by kvark** [comment]
Thanks @kainino0x !
I did another small pass, and changed the allowed sets to be "allowed multisets", which is what we are really trying to enforce here, mathematically speaking :)
As for the question of combining `input` + `constants` - I'm fine either way. I'm hesitating to do this because the idea of "let's combine all read-only usages together" is unfeasible in the extreme: the "storage-read" has to be separate anyway. So pushing it further wouldn't make it more elegant or simpler.

**Comment by kvark** [comment]
Not sure why CI isn't happy, it's passing for me locally

**Comment by Kangz** [comment]
As discussed offline, we should probably disallow `readonly-storage-texture` and `writeonly-storage-texture` from being used in combination because it will just encourage people to emulate `storage-texture` by adding two bindings (without having the guarantees for it). This restriction should be lifted with the `storage-texture` feature.

**Comment by kvark** [comment]
@Kangz this is already in this PR fwiw

File: spec/index.bs [hunk 12]
```diff
@@ -601,107 +601,122 @@ the user behind an `ArrayBuffer`, thus avoiding any data copies.
...
+We define <dfn dfn>subresource</dfn> to be either a whole buffer, or a [=texture subresource=].
+
+Some [=internal usage=]s are compatible with others.
+That means an implementation can consider multiple usages active at the same time.
+There is a limited set of <dfn dfn>allowed usage multiset</dfn>s, where each usage is compatible with the other:
+  - {[=internal usage/input=] | [=internal usage/constant=] | [=internal usage/storage-read=]} - in any combination
+  - {[=internal usage/storage=] | [=internal usage/storage-read=]} - in any combination
+  - {[=internal usage/storage-write=]} - used one or more times
+  - [=internal usage/attachment=] - singular usage
+  - {[=internal usage/attachment-read=] | [=internal usage/storage-read=] | [=internal usage/constant=]} - in any combination
```
Comment (kainino0x) [inline spec/index.bs:644]: As specified the allowed usage multisets should be infinite multisets, and I don't think that's clearly explained. E.g. the first one is {input &times; ∞, constant &times; ∞, storage-read &times; ∞} and the fourth one is {attachment &times; 1}.

Instead of defining the actual allowed multisets we can probably simply write them as expressions. Here's an incomplete suggestion:

```
<div algorithm=...>
    A [=usage multiset=] |U| is valid if it satisfies any of the following rules:
    - Each usage in |U| is [=internal usage/input=], [=internal usage/constant=], or [=internal usage/storage-read=].
    - Each usage in |U| is [=internal usage/storage=] or [=internal usage/storage-read=].
    - Each usage in |U| is [=internal usage/storage-write=].
    - Each usage in |U| is [=internal usage/attachment-read=], [=internal usage/storage-read=], or [=internal usage/constant=].
    - |U| contains exactly one element: [=internal usage/attachment=].
</div>
```

Alternative (but too wordy): could say something like "The set of usages in |U| is a subset of ..."

Comment (kvark) [reply spec/index.bs]: I liked multisets and having a set of them for the mathematical purity. But I agree that changing the wording to be more to the point ends up with a cleaner spec. We can save the math purity for implementation details ;)
So I removed the mention of "multiset" now, operating on the lists, and used the definition you suggested here.

File: spec/index.bs [hunk 13]
```diff
@@ -634,20 +634,16 @@ Each such <dfn dfn>texture subresource</dfn> is uniquely identified by a
...
-  - [=internal usage/attachment=] - singular usage
-  - {[=internal usage/attachment-read=] | [=internal usage/storage-read=] | [=internal usage/constant=]} - in any combination
-
-A multi-set differentiates between an `[=internal usage=]` included once versus multiple times.
-This difference is critical for [=internal usage/attachment=],
-which can't be used multiple times in the same render pass for the same [=texture subresource=].
-
-Enforcing that the usages are only combined into one of the [=allowed usage multiset=]s
+Some [=internal usage=]s are compatible with others. A [=subresource=] can be in a state
+that combines multiple usages together. We consider a list |U| to be
```
Comment (kainino0x) [inline spec/index.bs:638]: Should either wrap in an `algorithm` so the `|U|` var works, or remove the `| |`

File: spec/index.bs [hunk 14]
```diff
@@ -634,6 +634,7 @@ Each such <dfn dfn>texture subresource</dfn> is uniquely identified by a
 
 We define <dfn dfn>subresource</dfn> to be either a whole buffer, or a [=texture subresource=].
 
+<div algorithm>
 Some [=internal usage=]s are compatible with others. A [=subresource=] can be in a state
 that combines multiple usages together. We consider a list |U| to be 
 a <dfn dfn>compatible usage list</dfn> if (and only if) it satisfies any of the following rules:
```
Comment (kainino0x) [inline spec/index.bs:640]: I think this makes the algorithm title "compatible usage list", which is fine, just fyi

**Comment by grorg** [comment]
Discussed at the [2020-08-24 meeting](https://docs.google.com/document/d/1xKHeGs7d8KOjNPQuRIGYptXacITV9iVNhUTMoOGkTqM/edit#).
